File: algorithms/java/src/test/java/leetcode/LC387FirstUniqueCharacterInAString.java
package leetcode;

import org.junit.Test;

import static org.junit.Assert.assertEquals;

/**
 * https://leetcode.com/problems/first-unique-character-in-a-string/
 * #easy
 */
public class LC387FirstUniqueCharacterInAString {
    public class Solution {
        public int firstUniqChar(String s) {
            int size = 'z' - 'a' + 1;
            int[] count = new int[size];
            int[] index = new int[size];
            int length = s.length();
            for (int i = length - 1; 0 <= i; i--) {
                int key = s.charAt(i) - 'a';
                index[key] = i;
                count[key]++;
            }
            int min = -1;
            for (int i = 0; i < size; i++) {
                if (count[i] == 1) {
                    if (min == -1 || index[i] < min) {
                        min = index[i];
                    }
                }
            }
            return min;
        }
    }

    @Test
    public void test_leetcode() throws Exception {
        assertEquals(0, new Solution().firstUniqChar("leetcode"));
    }

    @Test
    public void test_loveleetcode() throws Exception {
        assertEquals(2, new Solution().firstUniqChar("loveleetcode"));
    }

    @Test
    public void test_empty() throws Exception {
        assertEquals(-1, new Solution().firstUniqChar(""));
    }
}

File: algorithms/code/leetcode/lc387_first_unique_character_in_a_string/lc387_first_unique_character_in_a_string.py
#!/usr/bin/env python3
# https://leetcode.com/problems/first-unique-character-in-a-string/

import unittest


class Solution:
    def firstUniqChar(self, s: str) -> int:
        size = ord('z') - ord('a') + 1
        count = [0] * size
        index = [0] * size
        length = len(s)
        for i in range(length - 1, -1, -1):
            key = ord(s[i]) - ord('a')
            index[key] = i
            count[key] += 1
        minimum = -1
        for i in range(size):
            if count[i] == 1 and (minimum == -1 or index[i] < minimum):
                minimum = index[i]
        return minimum


class TestCode(unittest.TestCase):
    def test_leetcode(self):
        self.assertEqual(0, Solution().firstUniqChar('leetcode'))

    def test_loveleetcode(self):
        self.assertEqual(2, Solution().firstUniqChar('loveleetcode'))

    def test_empty(self):
        self.assertEqual(-1, Solution().firstUniqChar(''))
